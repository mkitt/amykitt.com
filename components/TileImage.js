// @flow
import React from 'react'
import css from '../styles/css'

type Props = {
  allowableWidth: number,
  className?: string,
  src: string,
  naturalHeight: number,
  naturalWidth: number,
  onLoadSuccess?: Function | null,
  onLoadFailure?: Function | null,
}

const style = css({
  backgroundColor: '#7c7c7c',
})

export default class extends React.PureComponent {
  props: Props
  img: ?Image

  static defaultProps = {
    className: '',
    onLoadSuccess: null,
    onLoadFailure: null,
  }

  componentDidMount() {
    this.createLoader()
  }

  componentDidUpdate(prevProps: Props) {
    if (!this.img && prevProps.src !== this.props.src) {
      this.createLoader()
    }
  }

  componentWillUnmount() {
    this.disposeLoader()
  }

  onLoadSuccess = () => {
    if (typeof this.props.onLoadSuccess === 'function') {
      this.props.onLoadSuccess(this.img)
    }
    this.disposeLoader()
  }

  onLoadFailure = () => {
    this.disposeLoader()
    if (typeof this.props.onLoadFailure === 'function') {
      this.props.onLoadFailure()
    }
  }

  getDimensions() {
    const { allowableWidth, naturalWidth, naturalHeight } = this.props
    const ratio = naturalWidth ? naturalWidth / naturalHeight : NaN
    const maxCellHeight = 1200
    const widthConstrainedRelativeHeight = Math.round(allowableWidth * (1 / ratio))
    const height = Math.min(widthConstrainedRelativeHeight, naturalHeight, maxCellHeight)
    const width = Math.round(height * ratio)
    return { width, height }
  }

  createLoader() {
    this.disposeLoader()
    if (!this.props.src || !this.props.onLoadSuccess) { return }
    this.img = new Image()
    this.img.onload = this.onLoadSuccess
    this.img.onerror = this.onLoadFailure
    this.img.src = this.props.src
  }

  disposeLoader() {
    if (!this.img) { return }
    this.img.onload = null
    this.img.onerror = null
    this.img = null
  }

  render() {
    const { className, src } = this.props
    const { width, height } = this.getDimensions()
    return (
      <img
        alt={src.replace('-', ' ').replace('.jpg', '')}
        className={`${className || ''} ${style}`}
        src={src}
        width={width}
        height={height}
      />
    )
  }
}
