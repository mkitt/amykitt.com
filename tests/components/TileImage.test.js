import React from 'react'
import renderer from 'react-test-renderer'
import TileImage from '../../components/TileImage'

const noop = () => {}

describe('TileImage', () => {
  let component

  beforeEach(() => {
    component = (
      <TileImage
        allowableWidth={1440}
        className="TileImageClassName"
        src="/static/favicon.png"
        naturalHeight={600}
        naturalWidth={800}
        onLoadSuccess={noop}
        onLoadFailure={noop}
      />
    )
  })

  it('shows the correct snapshot for the TileImage component tree', () => {
    const tree = renderer.create(component).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
