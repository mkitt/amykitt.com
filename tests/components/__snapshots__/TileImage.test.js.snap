// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TileImage shows the correct snapshot for the TileImage component tree 1`] = `
<img
  alt="/static/favicon.png"
  className="TileImageClassName css-bagnb1"
  height={600}
  src="/static/favicon.png"
  width={800}
/>
`;
